### DISCUSSION TITLE
Open licensing discussion

### DISCUSSION THREAD

**Discussion by Torxed** [openzfs/zfs#12736] [General]
Hi.

This question/discussion is mainly targeted towards the OpenZFS summit and the inclusion of OpenZFS in open source community repositories and projects. I'm obviously not the [first one](https://ubuntu.com/blog/zfs-licensing-and-linux) to bring up [this topic](https://news.ycombinator.com/item?id=16377135) to this project, and I apologize for my own ignorance if that's the feeling - but I've tried to do my due diligence in order to avoid poking a bear for no good reason. And I hope that the somewhat recent change to *OpenZFS* would take a new grasp on this question.

I'm wondering if there's a conservation plan to phase out old code under the old licensing model? And would that be enough? and if so, are there any plans to move towards something more "mainstream" in terms of licensing, for instance MIT or GPL+BSD? Looking back at the [OpenZFS summits](https://openzfs.org/wiki/OpenZFS_Developer_Summit_2021), I can't see any talks devoted to these soft questions. It could be that I just missed or misunderstood the titles of the talks, but from a user and community standpoint, it's hard to get a clear grasp if there's a plan and if so what that plan is.

The reason why I create this question is because it keeps creeping up in different [projects](https://github.com/archlinux/archinstall/discussions/701) that I'm involved in, and have done so for a long long time. Not to mention my own love for ZFS and I use it through my own volition. But I'd like to be able to include it in more open source projects without having to worry about licensing conflicts and issues, specifically surrounding MIT and GPL licenses.

I get that this is a sensitive and loaded topic, so don't feel a need to respond if it's just to push this topic *(or me)* down for no reason other than just to do it. I welcome discussions and I'd love to learn more, not start or stir up something.

All the best, //Anton

**Comment by bghira** [comment]
anyone who wants to use it, can. the distributions that do, will benefit. the ones who keep insisting there's some license issue, won't. that's all there is to it, tbh.

phasing the old code out is infeasible, it would need to be a clean room reimplementation and the existing devs of OpenZFS wouldn't be allowed to work on it. 

in my *opinion*, the people who are disagreeing with Eben Moglen do so purely because of their own investments in things-that-aren't-ZFS.

**Reply by bghira** [reply]
the reason i say this about Eben Moglen, as he was the one behind the advice that led to Canonical's inclusion of ZFS:

> This situation was obviously problematic - Eben is clearly free to provide whatever legal opinion he holds to his clients, but his very public association with the FSF caused many people to assume that these positions were held by the FSF and the FSF were forced into the position of publicly stating that they disagreed with legal positions held by their general counsel. Attempts to mediate this failed, and Eben refused to commit to working with the FSF on avoiding this sort of situation in future[1].

the foundation's general counsel says that the license is to be interpreted in a certain way, and then non-lawyers decided he wasn't doing what they paid him for, and replaced him with other counsel that *did* say what they wanted them to say. but as their general counsel, he was pretty much the last word. even Matthew Garrett puts it succinctly:

> Eben has acted against the best interests of an organisation he publicly represented. He has threatened organisations and individuals who work to further free software. His actions are no longer to the benefit of the free software community and the free software community should cease associating with him.

their ideology was threatened by legal facts.

**Reply by Torxed** [reply]
We'll deviate from my original question, which was if there's a conservation plan to deal with this issue *(fictional or otherwise)*.

---

If the people voice concern over the chosen license, there surely must exist some basis for the concern? Or are they all completely factually inaccurate discussions? If you omit 10% at either end of the discussion, you're left with the middle part where people still voice concern. And it can't be that 100% of the people is either disagreeing *"just because" (being a nuisance)* - or disagreeing simply because other people are *(bandwagon mentality)*. Some of those people have to come from a place of good intention and facts and simply want a positive change.

So why the reluctance to do this change if it caters to more people? Is there a showstopper in BSD, GPL or MIT? If there is, then that's fine and that's the end of it. If there is not that would mean that the distributions that want to include it, could and would benefit without having to worry or think twice about it. And if the goal is to benefit more distributions of any kind, why not go down this path?

[BSD](https://opensource.org/licenses/BSD-3-Clause) and/or [GPL](https://www.gnu.org/licenses/gpl-3.0.html) for instance would solve a lot of the unknown people keep referring to, people wouldn't have anything to complain about and that would be the end of a long going discussion. And to the best of my knowledge, it wouldn't hurt or change anything surrounding OpenZFS.

> anyone who wants to use it, can.

Sadly not the case, we *(the common folk)* follow the creators verdicts and in this case GNU has officially and publicly said that [CDDL is not compatible with GPL](https://www.gnu.org/licenses/license-list.html#CDDL):

> it [CDDL] *is* incompatible with the GNU GPL

And [FSF also goes in to](https://www.fsf.org/licensing/zfs-and-linux) why CDDL can't work with *Linux* distributions.
And you could argue, that Richard Stallman and others just do this out of spite or because they don't like certain wordings, but I doubt that's the case when this many people have looked at it.

If the majority need this change to go through, why the reluctance is my question then I guess.

**Reply by bghira** [reply]
> If the people voice concern over the chosen license, there surely must exist some basis for the concern? Or are they all completely factually inaccurate discussions? If you omit 10% at either end of the discussion, you're left with the middle part where people still voice concern. And it can't be that 100% of the people is either disagreeing "just because" (being a nuisance) - or disagreeing simply because other people are (bandwagon mentality). Some of those people have to come from a place of good intention and facts and simply want a positive change.

there's different use cases for source code. anyone can compile the binary and load it into the kernel themselves. the _issue_ for Canonical is distributing a binary zfs.ko. at this point i should just suggest you do more research into the legal issues.

**Comment by mskarbek** [comment]
This was discussed many times, but since you insist.
You can't simply relicense OpenZFS - parts of the code are owned by Oracle, which bought Sun. During Sun times, developers not employed by Sun but working on OpenSolaris and its parts, including ZFS, had to assign copyrights to Sun which means that the whole code base was owned by Sun and Oracle would have to agree to relicence ZFS parts of it.
And throwing simple "BSD and/or GPL for instance would solve (something)" is far from the truth. Relicensing to GPL would cause license incompatibility with both FreeBSD and Illumos.
You can "technically" add BSD code to the Linux kernel. There are instances when kernel devs will require double licensing, though, and I don't have enough knowledge of such requirements in the file systems area to determine that switching to BSD would satisfy kernel devs.
So, first go to Larry Ellison and convince him that relicensing ZFS is a good business decision that will vastly profit Oracle in the future, then go back and start this discussion again.

**Reply by Torxed** [reply]
*"BSD and/or GPL for instance would solve (something)"* was taken slightly out of context, that thought belong to the argument that if weeding out old code was possible, I don't see how BSD *(there is a FreeBSD license that would be fine for the GPL world)* could hurt OpenZFS. And as stated, that would require a clean room and might not be feasible *(aka a complete re-write)*. And I get that, so I can't really argue against it nor is that my intention.

I appreciate you taking the time to respond and clarify a few things, I get that this topic might be re-posted every now and then. And my assumptions that OpenZFS would not be the same entity as the previous ZFS project, and that the source was re-written to some extent was perhaps misguided. My only followup would be that I don't have the necessary contacts to get in touch with Larry *(or anyone else legally involved in this matter)*, but hopefully the OpenZFS project would have. Therefor it would be better suited to direct that workload to the appropriate people - who hopefully had a dialogue when creating OpenZFS and the restructure/re-coding. But if there's no plan or road-map involving this type of work, or any intention or willingness to embark on such a steep project then that's fine too. And no matter how much the community would like to - nothing will happen on this front and that's it.

**Reply by mskarbek** [reply]
If such a dialog with Larry/Oracle/whatever were possible, it would have happened a long time ago. There is a lot of people that is/was waiting for it. :)

**Comment by jittygitty** [comment]
@Torxed I think I'm pretty familiar with this and I had "started" a very "detailed" explanation, unfortunately it was stated clearly it was a BRAINSTORM and heavily "under construction" "WIP" work in progress. But some people didn't have the patience to give me a chance to clean it up, make it clear and quickly jumped on me, and understandably I guess project maintainers felt they should close it to 'appease'.

You can read the deserted unfinished post here: https://github.com/openzfs/zfs/issues/11357#issuecomment-1115205182
Correction its at: https://github.com/openzfs/zfs/issues/13415

But to summarize for you, "silence" I think is prescribed by Oracle lawyers, which would be better than admitting they have need for the patent protections in CDDL which are completely missing from GPLv2. An ex-Sun employee, made some statement that Sun purposefully chose CDDL to prevent zfs in Linux (refuted by her supervisors and/or superiors, proof is on web) which unfortunately stuck in the mind of Linus and some other kernel devs but was simply not the real reason. The patent thing on the other hand, was a real issue.

Example, Microsoft uses Linux more than anything in their Azure etc, they could not file patent against Oracle if they used zfs under CDDL but they could if Oracle used GPL. Also there is a lot of confusion out there on this topic, people don't even realize that GPLv3 is actually just as "technically Incompatible" with GPLv2 as CDDL. That is why I even humored about doing a tongue in cheek commit to Linux kernel:
https://github.com/openzfs/zfs/issues/11357#issuecomment-1118003748

If denied one could respectfully ask them to fix their current typo to "export_symbol_GPLv2" since again GPLv3 is just as incompatible.

Ironically, the FSF would have something to gain out of such a patch I proposed, regardless of how much they might hate zfs or CDDL, and that is it would open up Linux Kernel modules to use GPLv3 and be able to use Linux GPL-only exports for these GPLv3 modules. Since Linus had made it clear I recall that kernel will never be GPLv3 itself, and FSF is trying hard to promote GPLv3, overall this would be a "win win" for everyone. The "intent" of Linus for copyleft would be preserved as well as Richard Stallman's original "intent" and many other projects under copyleft licenses but simply not using GPL "v2" would all "win" by simply moving to "Export_Symbol_Copyleft".

Linus and Kernel devs would have literally nothing to lose by such move, in fact, it would greatly bolster the power of the legal arguments in court when "proprietary" BINARY-only modules break the rules and don't release sources. Defending GPL in court against proprietary infringement would definitely benefit from such a move, any lawyer can confirm that, no need to explain all the reasons why over here.
